mets_adapter/mets_adapter/src/main/scala/uk/ac/wellcome/mets_adapter/services/MetsAdapterWorkerService.scala

```diff
@@ -14,43 +15,116 @@ import uk.ac.wellcome.mets_adapter.models._
 
 import scala.concurrent.Future
 
-case class StorageUpdate(space: String, bagId: String)
-
-case class MetsLocation(location: String)
-
+/** Processes SQS messages representing bag ingests on storage-service, and
+  *  publishes METS data for use in the pipeline.
+  *
+  *  Consists of the following stages:
+  *  - Retrieve the bag from storarge-service with the given ID
+  *  - Parse METS data from the bag
+  *  - Check whether the METS data is new / changed, and ignore it otherwise
+  *  - Publish the METS data to SNS
+  *  - Store the METS data so in future can tell if it's seen before
+  */
 class MetsAdapterWorkerService(
-  sqsStream: SQSStream[StorageUpdate],
-  snsMsgSender: SNSMessageSender,
+  msgStream: SQSStream[IngestUpdate],
+  msgSender: SNSMessageSender,
   bagRetriever: BagRetriever,
+  metsStore: MetsStore,
   concurrentConnections: Int = 6)(implicit ec: ExecutionContext)
-    extends Runnable {
+    extends Runnable
+    with Logging {
+
+  type Result[T] = Either[Throwable, T]
 
   val className = this.getClass.getSimpleName
 
   def run(): Future[Done] =
-    sqsStream.runStream(className, source => {
-      source
-        .via(retrieveBag)
-        .via(getMetsLocation)
-        .via(publishMetsLocation)
-    })
-
-  def retrieveBag: Flow[(SQSMessage, StorageUpdate), (SQSMessage, Bag), _] =
-    Flow[(SQSMessage, StorageUpdate)]
+    msgStream.runStream(
+      className,
+      source => {
+        source
+          .map { case (msg, update) => (Context(msg, update.bagId), update) }
+          .via(retrieveBag)
+          .via(parseMetsData)
+          .via(filterMetsData)
+          .via(publishMetsData)
+          .via(storeMetsData)
+          .map { case (Context(msg, _), _) => msg }
+      }
+    )
+
+  def retrieveBag =
```

Comment:
Maybe it's me just getting familiar with flows, but given they receive a type from the previous flow, could we add return types here?

Or do you think the `via` methods above imply that enough?

Whatever is more readable, I might just need to get more familiar with the syntax!

Reply:
I had included the full type signature for the flows, but at some point decided to remove them as a bit of a mouthful so added a lot of noise.

Without explicit types:

```scala
def parseMetsData =
  Flow[(Context, Option[Bag])]
  ...
```

With explicit types:

```scala
def parseMetsData: Flow[(Context, Option[Bag]), (Context, Option[MetsData]), NotUsed] =
  Flow[(Context, Option[Bag])]
  ...
```

I personally think it's pretty readable without, as it is easy to see the output from the input to the next flow, but happy to be explicit if people would prefer.